### PR TITLE
Queue auto-tuning part 1

### PR DESCRIPTION
**Pull request by rguo-aws**
*Issue #, if available:*

*Description of changes:*
This PR is the the part 1 of my PR for queue auto-tuning changes. This part is to add queue rejection RCA into RCA graph. Queue rejection will be one of the input for auto tuning policy. 

*Tests:*
Tested on docker.

*Code coverage percentage for this patch:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/ElasticSearchAnalysisGraph.java [hunk 1]
```diff
@@ -124,22 +130,31 @@ public void construct() {
     highCpuRca.addTag(TAG_LOCUS, LOCUS_DATA_MASTER_NODE);
     highCpuRca.addAllUpstreams(Collections.singletonList(cpuUtilizationGroupByOperation));
 
-    Rca<ResourceFlowUnit> hotJVMNodeRca = new HotNodeRca(12, highHeapUsageOldGenRca,
-            highHeapUsageYoungGenRca, highCpuRca);
-    hotJVMNodeRca.addTag(TAG_LOCUS, LOCUS_DATA_MASTER_NODE);
-    hotJVMNodeRca.addAllUpstreams(
-            Arrays.asList(highHeapUsageOldGenRca, highHeapUsageYoungGenRca, highCpuRca));
+    Rca<ResourceFlowUnit> queueRejectionRca = new QueueRejectionRca(12, threadPool_RejectedReqs);
```
Comment (yojs) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/ElasticSearchAnalysisGraph.java:133]: Can we extract the RCA into its own method ?

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/ElasticSearchAnalysisGraph.java [hunk 2]
```diff
@@ -124,22 +130,31 @@ public void construct() {
...
+    queueRejectionRca.addTag(TAG_LOCUS, LOCUS_DATA_MASTER_NODE);
+    queueRejectionRca.addAllUpstreams(Collections.singletonList(threadPool_RejectedReqs));
+
+    Rca<ResourceFlowUnit> hotNodeRca = new HotNodeRca(12, highHeapUsageOldGenRca,
```
Comment (yojs) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/ElasticSearchAnalysisGraph.java:137]: A bunch of the RCAs take `12` as the constant. Can we create a variable with appropriate name and use it for readability ?

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/queue/QueueRejectionRca.java [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.scheduler.FlowUnitOperationArgWrapper;
+import com.google.common.annotations.VisibleForTesting;
+import java.time.Clock;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class QueueRejectionRca extends Rca<ResourceFlowUnit> {
```
Comment (yojs) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/queue/QueueRejectionRca.java:42]: Can we add a javadoc describing how this RCA operates semantically  ?

Comment (rguo-aws) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/queue/QueueRejectionRca.java:47]: done

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/ElasticSearchAnalysisGraph.java (hunk 1 above)
Comment (rguo-aws) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/ElasticSearchAnalysisGraph.java]: done

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/queue/QueueRejectionRca.java (hunk 3 above)
Comment (rguo-aws) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/queue/QueueRejectionRca.java:47]: done

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/ElasticSearchAnalysisGraph.java (hunk 2 above)
Comment (rguo-aws) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/ElasticSearchAnalysisGraph.java]: done

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/queue/QueueRejectionRca.java [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+        context = new ResourceContext(Resources.State.HEALTHY);
+      }
+      else {
+        context = new ResourceContext(Resources.State.UNHEALTHY);
+        summary = new HotResourceSummary(this.resourceType,
+            0, threadpoolSummaries.size(), REJECTION_TIME_PERIOD_IN_SECONDS);
+        threadpoolSummaries.sort(
+            (TopConsumerSummary t1, TopConsumerSummary t2) -> Double.compare(t2.getValue(), t1.getValue())
+        );
+        summary.addNestedSummaryList(threadpoolSummaries);
```
Comment (khushbr) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/queue/QueueRejectionRca.java:106]: Wondering if we should have separate top consumers for WRITE and SEARCH.

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/ElasticSearchAnalysisGraph.java [hunk 5]
```diff
@@ -82,68 +85,115 @@
...
-    Metric gcEvent = new GC_Collection_Event(5);
-    Metric heapMax = new Heap_Max(5);
-    Metric gc_Collection_Time = new GC_Collection_Time(5);
-    Metric cpuUtilizationGroupByOperation = new AggregateMetric(1, CPU_Utilization.NAME,
-            AggregateFunction.SUM,
-            MetricsDB.AVG, CommonDimension.OPERATION.toString());
+    //construct node level metrics and RCAs. HotNodeRCA will subscribe to all node level RCAs
+    Rca<ResourceFlowUnit> hotNodeRca = constructNodeLevelRcas();
+    //construct cluster level RCAs
+    constructClusterLevelRcas(hotNodeRca);
```
[the remark this answers is not in the document]
Comment (rguo-aws) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/ElasticSearchAnalysisGraph.java:96]: agree. a TODO has been added.

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/cluster/GenericClusterRca.java [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class GenericClusterRca extends Rca<ResourceFlowUnit> {
+  private static final Logger LOG = LogManager.getLogger(GenericClusterRca.class);
+  private static final int UNHEALTHY_FLOWUNIT_THRESHOLD = 3;
```
[the remark this answers is not in the document]
Comment (rguo-aws) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/cluster/GenericClusterRca.java:47]: This is more like a hard coded parameter which doesn't need to be changed in runtime. Will consider adding it to conf if we want to tune it in the future. but I would like to leave it there at this moment

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/cluster/GenericClusterRca.java [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class GenericClusterRca extends Rca<ResourceFlowUnit> {
```
[the remark this answers is not in the document]
Comment (rguo-aws) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/cluster/GenericClusterRca.java:45]: done

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/queue/QueueRejectionRca.java (hunk 4 above)
Comment (rguo-aws) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/queue/QueueRejectionRca.java:106]: As per our discussion, this is how we planned in the design doc

**Pull request closed without merging**